### PR TITLE
tracing: add baggage methods to Tracing::Span

### PR DESCRIPTION
**Pull request by matthewgrossman**
Commit Message: tracing: add baggage methods to Tracing::Span
Additional Description: references https://github.com/envoyproxy/envoy/issues/11622: We want an envoy filter to be able to read information off of the [OT baggage](https://opentracing.io/docs/overview/tags-logs-baggage/) to guide some routing decisions. First step is exposing the functionality of the existing spans/tracers so we can access this baggage dict within the filters. Once this functionality is added to the `Tracing::Span`, we can access with the [baggage info inside a filter](https://github.com/envoyproxy/envoy/blob/b2610c84aeb1f75c804d67effcb40592d790e0f1/include/envoy/http/filter.h#L135)
Risk Level: Low: Small optional feature
Testing: Unit testing is added for all derived classes of `Tracing::Span`. All of the classes are noops except `OpenTracingSpan` ([test](https://github.com/envoyproxy/envoy/pull/12260/files#diff-290f0f256bca554083d2fa92593d2b06)) and LightStep ([test](https://github.com/envoyproxy/envoy/pull/12260/files#diff-ae21654f7c0e5b45467220040275d85f))
Docs Changes: N/A for now, should something about baggage be added to [tracing.rst](https://github.com/matthewgrossman/envoy/blob/e75f095915e25957729893bfd13f1226820a8474/docs/root/intro/arch_overview/observability/tracing.rst)?
Release Notes: think we need a note here?

### CODE REVIEW

File: source/extensions/tracers/common/ot/opentracing_driver_impl.cc [hunk 1]
```diff
@@ -102,6 +102,12 @@ void OpenTracingSpan::log(SystemTime timestamp, const std::string& event) {
   finish_options_.log_records.emplace_back(std::move(record));
 }
 
+void OpenTracingSpan::setBaggage(const std::string& key, const std::string& value) {
+  span_->SetBaggageItem(key, value);
+}
+
+std::string OpenTracingSpan::getBaggage(const std::string& key) { return span_->BaggageItem(key); }
```
Comment (matthewgrossman) [inline source/extensions/tracers/common/ot/opentracing_driver_impl.cc:109]: these two methods really just delegate to the underlying Span impl, lightstep's found here

https://github.com/lightstep/lightstep-tracer-cpp/blob/8b7f871242a5d9f514a6eddf0ae51c15749d5ce7/src/tracer/span.cpp#L127-L159

File: include/envoy/tracing/http_tracer.h [hunk 2]
```diff
@@ -158,6 +158,20 @@ class Span {
...
+   * @return the baggage's value for the given input key
+   */
+  virtual std::string getBaggage(const std::string& key) PURE;
+
+  /**
+   * Set a key/value pair in the current span's baggage
+   * @param key baggage key
+   * @param key baggage value
+   */
+  virtual void setBaggage(const std::string& key, const std::string& value) PURE;
```
Comment (kyessenov) [inline include/envoy/tracing/http_tracer.h:174]: What's the difference with setTag?

Comment (matthewgrossman) [reply include/envoy/tracing/http_tracer.h:174]: baggage is persisted in the `SpanContext`, which means that any subsequent spans can access data in the baggage (it gets injected/extracted across process boundaries). This is useful when you need to embed information that is available for the entirety of the request

more reading: https://opentracing.io/docs/overview/tags-logs-baggage/, https://github.com/opentracing/specification/blob/master/specification.md#set-a-baggage-item

File: source/extensions/tracers/xray/tracer.h [hunk 3]
```diff
@@ -132,6 +132,9 @@ class Span : public Tracing::Span, Logger::Loggable<Logger::Id::config> {
    */
   void log(Envoy::SystemTime, const std::string&) override {}
 
+  void setBaggage(const std::string&, const std::string&) override {}
+  std::string getBaggage(const std::string&) override { return std::string(); }
```
Comment (matthewgrossman) [inline source/extensions/tracers/xray/tracer.h:136]: implemented in the header for xray simply because it seems thats the convention for this tracer

Comment (matthewgrossman) [reply source/extensions/tracers/xray/tracer.h:136]: (as opposed to the .cc)

File: source/extensions/tracers/opencensus/opencensus_tracer_impl.cc [hunk 4]
```diff
@@ -192,6 +194,9 @@ void Span::log(SystemTime /*timestamp*/, const std::string& event) {
   // timestamp is ignored.
   span_.AddAnnotation(event);
 }
+void Span::setBaggage(const std::string&, const std::string&) {}
+
+std::string Span::getBaggage(const std::string&) { return std::string(); }
```
Comment (LisaLudique) [inline source/extensions/tracers/opencensus/opencensus_tracer_impl.cc:199]: consider logging a warning here that this tracer doesn't support baggage (and likewise for the other tracers that do not support it)

Comment (LisaLudique) [reply source/extensions/tracers/opencensus/opencensus_tracer_impl.cc:199]: you could even log an error saying baggage isn't supported similar to https://github.com/envoyproxy/envoy/blob/68df9d17348fea587784da55d6e6aa8c3b06d743/source/extensions/tracers/opencensus/opencensus_tracer_impl.cc#L358

Comment (mattklein123) [reply source/extensions/tracers/opencensus/opencensus_tracer_impl.cc:199]: A warning isn't safe in the data path. I would either do nothing and leave a TODO or have a stat for dropped baggage. Same elsewhere.

File: include/envoy/tracing/http_tracer.h [hunk 5]
```diff
@@ -158,6 +158,20 @@ class Span {
    * @param sampled whether the span and any subsequent child spans should be sampled
    */
   virtual void setSampled(bool sampled) PURE;
+
+  /**
+   * Retrieve a key's value from the current span's baggage
+   * @param key baggage key
+   * @return the baggage's value for the given input key
+   */
+  virtual std::string getBaggage(const std::string& key) PURE;
```
[the remark this answers is not in the document]
Comment (matthewgrossman) [reply include/envoy/tracing/http_tracer.h]: done

File: include/envoy/tracing/http_tracer.h [hunk 6]
```diff
@@ -158,6 +158,20 @@ class Span {
...
+
+  /**
+   * Set a key/value pair in the current span's baggage
```
[the remark this answers is not in the document]
Comment (matthewgrossman) [reply include/envoy/tracing/http_tracer.h]: done in https://github.com/envoyproxy/envoy/pull/12260/commits/4abed0b3ea2f6b0ad40a009fafaa2e3a5bbf31cb

File: source/extensions/tracers/opencensus/opencensus_tracer_impl.cc (hunk 4 above)
Comment (matthewgrossman) [reply source/extensions/tracers/opencensus/opencensus_tracer_impl.cc]: I left comments for the spans that don't support baggage (opencensus, xray) and a TODO for zipkin

**Comment by matthewgrossman** [comment]
tuned up the PR with tests/comments/todos, tried to get the description to match the template as well. Might be worth reading that again since it's likely changed since last review.

Do we need to express anywhere else that not all implementations of `Tracing::Span` currently support the baggage methods? Maybe that's what would be valuable somewhere in some documentation

**Comment by LisaLudique** [comment]
tests look good!

I would suggest adding a section on baggage and delineating the level of support for each tracer in https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/observability/tracing

File: source/extensions/tracers/zipkin/zipkin_tracer_impl.cc [hunk 7]
```diff
@@ -35,6 +35,10 @@ void ZipkinSpan::log(SystemTime timestamp, const std::string& event) {
   span_.log(timestamp, event);
 }
 
+void ZipkinSpan::setBaggage(absl::string_view, absl::string_view) {}
```
Comment (LisaLudique) [inline source/extensions/tracers/zipkin/zipkin_tracer_impl.cc:38]: nit: can you add the TODO for this being unimplemented here as well?

File: test/extensions/tracers/xray/tracer_test.cc [hunk 8]
```diff
@@ -100,6 +100,14 @@ TEST_F(XRayTracerTest, NonSampledSpansNotSerialized) {
   span->finishSpan();
 }
 
+TEST_F(XRayTracerTest, BaggageNotImplemented) {
+  Tracer tracer{"" /*span name*/, std::move(broker_), server_.timeSource()};
+  auto span = tracer.createNonSampledSpan();
+  span->setBaggage("baggage_key", "baggage_value");
+  span->finishSpan();
+  ASSERT_EQ("", span->getBaggage("baggage_key"));
```
Comment (LisaLudique) [inline test/extensions/tracers/xray/tracer_test.cc:108]: nit: can you add a comment saying this isn't supported here as well to clarify testing for the empty value

File: source/extensions/tracers/zipkin/zipkin_tracer_impl.cc (hunk 7 above)
Comment (matthewgrossman) [reply source/extensions/tracers/zipkin/zipkin_tracer_impl.cc:39]: fixed in https://github.com/envoyproxy/envoy/pull/12260/commits/05dbaa50f5cc45f851bf881120f849f2d4e5e80f

File: test/extensions/tracers/xray/tracer_test.cc (hunk 8 above)
Comment (matthewgrossman) [reply test/extensions/tracers/xray/tracer_test.cc:110]: fixed in https://github.com/envoyproxy/envoy/pull/12260/commits/05dbaa50f5cc45f851bf881120f849f2d4e5e80f

File: source/extensions/tracers/opencensus/opencensus_tracer_impl.cc [hunk 9]
```diff
@@ -192,6 +196,9 @@ void Span::log(SystemTime /*timestamp*/, const std::string& event) {
   // timestamp is ignored.
   span_.AddAnnotation(event);
 }
+void Span::setBaggage(absl::string_view, absl::string_view) {}
+
+std::string Span::getBaggage(absl::string_view) { return std::string(); }
```
Comment (mattklein123) [inline source/extensions/tracers/opencensus/opencensus_tracer_impl.cc:201]: nit: just inline this above.

**Comment by matthewgrossman** [comment]
thanks for the help in review!

**Pull request merged**